Bug 1147071 - Use sqlite cipher

wesj opened the pull request.

st3fan commented on Cartfile.resolved:
```diff
@@ -1,11 +1,12 @@
 github "Masonry/Snap" "0.11.0"
-github "DaveWoodCom/XCGLogger" "29b77cc64147aa09fe233ad2c90007fb4cd1fbf5"
+github "DaveWoodCom/XCGLogger" "74390d672ed2256360ea48f289286718afc9efe6"
+github "sqlcipher/sqlcipher" "v3.3.0"
```
Can't we use the SQLCipher.framework that is part of SQLite.swift?

wesj replied:
I'm a bit nervous to use something that's dependent on another framework for SwiftData. Having SQLite.swift in the tree makes things... confusing here.

rnewman commented:
Can we have an easy method for disabling this during development? "No key, please."

Aside from that, some really important questions around lifecycle:

When a user backs up and restores their phone, does the keychain always come with it? I really really want to avoid users saying "hey, all of my data is gone, how do i get it back?"

What's the keychain access flag for the key used here? (<https://developer.apple.com/library/ios/documentation/Security/Reference/keychainservices/#//apple_ref/doc/constant_group/Keychain_Item_Accessibility_Constants>)

This answers the question: if the app is inited in the background with the phone locked (e.g., in response to a notification, or for a scheduled sync), can we get the key to access the DB?

rnewman commented:
Some comments on the commits. Higher-level comments:

• I think this would feel a little safer if we did it in baby steps: e.g., switch sqlite implementations without adding encryption, then extend and upstream KeychainWrapper, then use both.

• I think we need a firmer understanding of backup/iCloud/restore/extension access to Keychain before we commit. @st3fan probably has most of this in his head already, so perhaps chat with him?

• And on a related note: we can likely reduce our risk here by putting passwords in a separate database, rather than just a table in BrowserDB.

Not only does that avoid the need for us to encrypt the whole world, and the need to encrypt the whole world with the same key, but it also will make debugging easier and more polite — you won't accidentally send your passwords to a developer when asking them to debug your history!

On top of that, we get the usual advantages of failures not necessarily destroying both sets of data, avoidance of potential contention, etc. at the cost of a little more memory.

Thoughts?

st3fan commented:
lttm

fxios-buildasaur commented:
Result of integration 1
Integration took 2 minutes and 51 seconds.
30 build errors: build-errors

wesj commented:
lttm

fxios-buildasaur commented:
Result of integration 2
Integration took 2 minutes and 51 seconds.
30 build errors: build-errors

wesj commented:
No one seems to be able to tell me how to actually see these errors. All I can do is guess and push. Can we either turn this off or find something better st3fan?

rnewman commented:
https://ci.fxios.allizom.org/xcode/bots/failed/53EA4E0

wesj commented:
Thanks. It looks like this isn't finding a sqlite to link against. Is it possible that the builder isn't updating Carthage? Thinking....

wesj commented:
lttm

fxios-buildasaur commented:
Result of integration 3
Integration took 5 minutes and 49 seconds.
12 build errors: build-errors

wesj commented:
lttm

fxios-buildasaur commented:
Result of integration 4
Integration took 12 minutes and 28 seconds.
8 build errors: build-errors

wesj commented:
lttm

wesj commented:
lttm

wesj commented:
lttm

fxios-buildasaur commented:
Result of integration 5
Integration took 10 minutes and 20 seconds.
8 build errors: build-errors

wesj commented:
lttm

fxios-buildasaur commented:
Result of integration 6
Integration took 1 minute and 41 seconds.
7 build errors: build-errors

fxios-buildasaur commented:
Result of integration 7
Integration took 6 minutes and 2 seconds.
9 build errors: build-errors

fxios-buildasaur commented:
Result of integration 8
Integration took 5 minutes and 28 seconds.
9 build errors: build-errors

st3fan commented:
lttm

fxios-buildasaur commented:
Result of integration 9
Integration took 2 minutes and 0 second.
2 build errors: build-errors

fxios-buildasaur commented:
Result of integration 10
Integration took 1 minute and 5 seconds.
1 build errors: build-errors

st3fan commented:
lttm

fxios-buildasaur commented:
Result of integration 1
Integration took 2 minutes and 36 seconds.
2 build errors: build-errors

fxios-buildasaur commented:
Result of integration 3
Integration took 2 minutes and 29 seconds.
2 build errors: build-errors

fxios-buildasaur commented:
Result of integration 4
Integration took 6 minutes and 16 seconds.
9 build errors: build-errors

fxios-buildasaur commented:
Result of integration 5
Integration took 2 minutes and 9 seconds.
2 build errors: build-errors

fxios-buildasaur commented:
Result of integration 6
Integration took 1 minute and 55 seconds.
2 build errors: build-errors

fxios-buildasaur commented:
Result of integration 8
Integration took 58 seconds.
1 build errors: build-errors

wesj commented:
Closing. I rebased this in a different branch to fix the buildbot errors. They're all working now (I manually imported sqlcipher and changed some of its build settings). I'm going to land the other branch since this was r+ until we hit those problems.
